waggle-dance-extensions/src/main/java/com/hotels/bdp/waggledance/extensions/client/ratelimit/RateLimitingInvocationHandler.java

```diff
@@ -0,0 +1,121 @@
+  private String metastoreName;
+  private CloseableThriftHiveMetastoreIface client;
+  private String user = UNKNOWN_USER;
+
+  private final BucketService bucketService;
+  private final BucketKeyGenerator bucketKeyGenerator;
+
+  public RateLimitingInvocationHandler(
+      CloseableThriftHiveMetastoreIface client,
+      String metastoreName,
+      BucketService bucketService, BucketKeyGenerator bucketKeyGenerator) {
+    this.client = client;
+    this.metastoreName = metastoreName;
+    this.bucketService = bucketService;
+    this.bucketKeyGenerator = bucketKeyGenerator;
+  }
+
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    if (method.getName().equals("set_ugi")) {
+      user = (String) args[0];
+    }
+    if (isIgnoredMethod(method)) {
+      return doRealCall(client, method, args);
+    } else {
+      return doRateLimitCall(client, method, args);
+    }
+  }
+
+  private Object doRateLimitCall(CloseableThriftHiveMetastoreIface client, Method method, Object[] args)
+    throws IllegalAccessException, Throwable {
+    boolean proceedWithCall = proceedWithCall(method);
+    if (proceedWithCall) {
+      return doRealCall(client, method, args);
+    } else {
+      log.info("User '{}' made too many requests.", user);
+      // HTTP status would be 429, so using same for Thrift.
+      throw new WaggleDanceServerException("[STATUS=429] Too many requests.");
+    }
+  }
+
+  private boolean proceedWithCall(Method method) {
+    try {
+      Bucket bucket = bucketService.getBucket(bucketKeyGenerator.generateKey(user));
+      ConsumptionProbe probe = bucket.tryConsumeAndReturnRemaining(1);
+      log
+          .info("RateLimitCall:[User:{}, method:{}, source_ip:{}, tokens_remaining:{}, metastoreName:{}]", user,
+              method.getName(), HMSHandler.getThreadLocalIpAddress(), probe.getRemainingTokens(), metastoreName);
+      return probe.isConsumed();
+    } catch (Exception e) {
+      if (log.isDebugEnabled()) {
+        log.error("Error while processing rate limit for: User:{}, method:{}", user, method.getName(), e);
+      } else {
+        log
+            .error("Error while processing rate limit for: User:{}, method:{}, message:{}", user, method.getName(),
+                e.getMessage());
+      }
+      return true;
+    }
+  }
+
+  private Object doRealCall(CloseableThriftHiveMetastoreIface client, Method method, Object[] args)
+    throws IllegalAccessException, Throwable {
+    try {
+      return method.invoke(client, args);
+    } catch (InvocationTargetException e) {
+      Throwable realException = e.getTargetException();
+      throw realException;
+    }
+  }
+
+  /**
+   * Ignore some methods that are not "real" metastore calls or should not count towards a rate limit.
+   * 
+   * @param method
+   * @return true if the method should be ignored for rate limiting purposes.
+   */
+  private boolean isIgnoredMethod(Method method) {
```

Comment:
If you only need the method name, you should only pass that to the method. The less this method knows about the `Method` object the better (law of demeter).